src/data_struct/revision_vec.rs

```diff
@@ -0,0 +1,377 @@
+use std::collections::VecDeque;
+
+/// a `RevisionVec` stores for each entry a linked list of versions.
+/// The entry versions are stored in reverse insertion order:
+///
+/// Vec [
+///     0: key -> a" -> a' -> a
+///     1: key -> b
+///     2: key -> c' -> c
+/// ]
+///
+/// Insertions are only allowed at the front of the linked list.
+/// Deletions can only happen at the end of the linked list.
+///
+/// This guarantees that the entry versions are always ordered.
+#[derive(Default, Debug, PartialEq, Eq)]
+// TODO does index matter for Eq compare?
+pub struct RevisionVec<K, T> {
+    chains: Vec<(K, RevisionList<T>)>,
+}
+
+impl<K, T> RevisionVec<K, T> {
+    pub fn new() -> Self {
+        Self { chains: Vec::new() }
+    }
+
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            chains: Vec::with_capacity(capacity),
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.chains.iter().map(|(_, chain)| chain.len()).sum()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn nb_chains(&self) -> usize {
+        self.chains.len()
+    }
+
+    /// Creates and insert a new chain with a single value.
+    /// /!\ Adding multiple chains with the same key will corrupt the data
+    /// structure.
```

Comment:
I wonder if we get the correct concept here.

Reply:
`The check of the keys uniqueness is left to the user of this structure`?